VisualPinball.Unity/VisualPinball.Unity/Game/CameraClipPlane.cs

```diff
@@ -0,0 +1,142 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+using System; 
+using Unity.Mathematics;
+using UnityEngine;
+
+
+namespace VisualPinball.Unity
+{
+	[ExecuteAlways]
+	public class CameraClipPlane : MonoBehaviour
+	{
+		
+		[NonSerialized]
+		public Camera Camera;
+
+		//Stores the table bounds. 
+		private Bounds tableBounds;
+		private Vector3 previousCameraPosition = Vector3.zero;
+
+		private void OnEnable()
+		{
+			SetClipPlanes(0.001f, 9f); 
+		}
+
+		private void Update()
+		{
+			if(!GetCameraComponent())
+			{
+				return;
+			}
+
+			if(previousCameraPosition != Camera.transform.position)
+			{
+				UpdateClipPlanes();
+				previousCameraPosition = Camera.transform.position; 
+			}
+		}
+
+		/// <summary>
+		/// Updates the camera clipping planes based on the table bounds.
+		/// </summary>
+		public bool UpdateClipPlanes()
+		{
+
+			//Early out if no table is selected.  
+			if(!TableSelector.Instance.HasSelectedTable)
+			{
+				return false;
+			}
+
+			//Early out if we don't have a camera to work on. 
+			if(!GetCameraComponent())
+			{
+				return false;
+			}
+
+			//Get selected table reference. 
+			var table = TableSelector.Instance.SelectedTable; 
+
+			if(Application.isPlaying)
+			{
+				tableBounds = table._tableBounds;  //When playing at runtime, get the stored table bounds value instead of calculating. 
+			}
+			else
+			{
+				tableBounds = table.GetTableBounds();  //When in editor, calculate the bounds in case things have changed.  
+			}
+
+			var trans = Camera.transform;
+
+			var cameraPos = trans.position; // camera position. 
+			var sphereExtent = tableBounds.extents.magnitude; //sphere radius of the bounds. 
+			float cameraToCenter = Vector3.Distance(cameraPos, tableBounds.center);
+			var nearPlane = math.max(cameraPos.magnitude - sphereExtent, 0.001f); //Assign initial near plane used when camera is not in the sphere. 
+			var farPlane = math.max(1f, nearPlane + sphereExtent * 2f); //initial far bounds 
+
+			if(cameraToCenter < sphereExtent)
+			{
+				nearPlane = 0.01f; //camera is in the bounds so drop the near plane to very low. 
+				farPlane = math.max(0.01f, sphereExtent + Vector3.Distance(cameraPos, tableBounds.center)); //set far plane to delta between camera and furthest bound. 
+
+			}
+
+			SetClipPlanes(nearPlane, farPlane); 
+
+			return true; 
+		}
+
+		/// <summary>
+		/// Gets the currently active or attached camera component. 
+		/// </summary>
+		/// <returns>False when no camera can be found.</returns>
+		private bool GetCameraComponent()
```

Comment:
Maybe I have done too much Java in the past, but "Get" methods usually indicate a quick operation (e.g. returning an already set member). Here you're doing `GetComponent()` which is more expensive, so I'd rather call it `RetrieveCameraComponent()`.

Reply:
Sounds good.